### PR TITLE
Add Snom UXMC D7C firmware entry

### PR DESCRIPTION
**Pull request by BnLG**
Hello,

This PR adds Snom UXMC D7C firmware entry, making it available through Wazo portal.
https://service.snom.com/display/wiki/D7C
https://service.snom.com/display/wiki/UXM+-+D7C+-+V1.1.1

It also updates the `firmware_uxm` option in the Snom plugin template, depending on what has been downloaded.

Finally, this PR updates Snom firmware links to HTTPS, as their HTTP links do not respond most of the time.
No issue with HTTPS.

Ref : Wazo ticket #5638.

Thank you :+1:

### CODE REVIEW

**Comment by BnLG** [comment]
Thanks to this PR we will have both UXM and UXMC firmwares available.
Both are to be provided thanks to the `<firmware_uxm>` option.

I would then have liked to update the plugin with something like this :
```
{% if os.path.exists('firmware/snomUXM-2.1.1.bin') %}
<firmware_uxm>http://{{ ip }}:{{ http_port }}/firmware/snomUXM-2.1.1.bin</firmware_uxm>
{% else %}
<firmware_uxm>http://{{ ip }}:{{ http_port }}/firmware/snomD7C-1.1.1-r.bin</firmware_uxm>
{% endif %}
````

However, `wazo-provd` returns the following :

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/provd/app.py", line 278, in _dev_configure
    except Exception:
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/common.py", line 327, in configure
    self._add_fkeys(raw_config, model)
  File "/usr/lib/python2.7/dist-packages/provd/plugins.py", line 471, in dump
    template.stream(context).dump(tmp_filename, encoding, errors)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1227, in dump
    else:
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1223, in <genexpr>
    else:
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1271, in __next__
    
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1045, in generate
    
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 780, in handle_exception
    
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/templates/D735.tpl", line 1, in top-level template code
    {% extends 'base.tpl' -%}
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/templates/base.tpl", line 35, in top-level template code
    {% if os.path.exists('/tmp/dummy.csv') -%}
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 430, in getattr
    except AttributeError:
UndefinedError: 'os' is undefined
```

Could you then help me making `os.path.exists` to properly work in plugins please ?

Many thanks 👍

**Comment by BnLG** [comment]
Thanks to this PR we will have both UXM and UXMC firmwares available.
Both are to be provided thanks to the `<firmware_uxm>` option.

I would then have liked to update the plugin with something like this :
```
{% if os.path.exists('firmware/snomUXM-2.1.1.bin') %}
<firmware_uxm>http://{{ ip }}:{{ http_port }}/firmware/snomUXM-2.1.1.bin</firmware_uxm>
{% else %}
<firmware_uxm>http://{{ ip }}:{{ http_port }}/firmware/snomD7C-1.1.1-r.bin</firmware_uxm>
{% endif %}
````

However, `wazo-provd` returns the following :

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/provd/app.py", line 278, in _dev_configure
    except Exception:
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/common.py", line 327, in configure
    self._add_fkeys(raw_config, model)
  File "/usr/lib/python2.7/dist-packages/provd/plugins.py", line 471, in dump
    template.stream(context).dump(tmp_filename, encoding, errors)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1227, in dump
    else:
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1223, in <genexpr>
    else:
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1271, in __next__
    
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 1045, in generate
    
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 780, in handle_exception
    
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/templates/D735.tpl", line 1, in top-level template code
    {% extends 'base.tpl' -%}
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/templates/base.tpl", line 35, in top-level template code
    {% if os.path.exists('/tmp/dummy.csv') -%}
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 430, in getattr
    except AttributeError:
UndefinedError: 'os' is undefined
```

Could you then help me making `os.path.exists` to properly work in plugins please ?

Many thanks 👍

**Comment by BnLG** [comment]
This will need https://github.com/wazo-platform/wazo-provd-plugins/pull/136 first.
Thx 👍

**Comment by BnLG** [comment]
This will need https://github.com/wazo-platform/wazo-provd/pull/50 first.
Thx 👍

File: plugins/wazo-snom/common/common.py [hunk 1]
```diff
@@ -156,6 +157,14 @@ def __init__(self, app, plugin_dir, gen_cfg, spec_cfg):
 
     http_dev_info_extractor = BaseSnomHTTPDeviceInfoExtractor()
 
+    def _add_uxm_firmware(self, raw_config):
+        f = glob.glob(self._tftpboot_dir + u'/firmware/snomD7C-*.bin') + glob.glob(self._tftpboot_dir + u'/firmware/snomUXM-*.bin')
```
Comment (bloom1) [inline plugins/wazo-snom/common/common.py:161]: It is unsafe to use `+` to join a path: use `os.path.join` (as on line 171) to concatenate a filepath :smile:

File: plugins/wazo-snom/common/common.py [hunk 2]
```diff
@@ -156,6 +157,14 @@ def __init__(self, app, plugin_dir, gen_cfg, spec_cfg):
...
+        if len(f):
```
Comment (bloom1) [inline plugins/wazo-snom/common/common.py:162]: You can just do `if f:` - an empty list is false-y

File: plugins/wazo-snom/common/common.py [hunk 3]
```diff
@@ -156,6 +157,14 @@ def __init__(self, app, plugin_dir, gen_cfg, spec_cfg):
...
+            if re.match(r"^.*\/snomUXM-.*.bin$", max(f, key=os.path.getmtime)):
+                raw_config[u'XX_uxm_firmware'] = 'uxm'
+            if re.match(r"^.*\/snomD7C-.*.bin$", max(f, key=os.path.getmtime)):
```
Comment (bloom1) [inline plugins/wazo-snom/common/common.py:165]: I think that having the `glob` in two different lists would avoid the need to do a regex

**Review by bloom1 — CHANGES_REQUESTED** [review]
Great job! 
I have a few points though :smile: 

Thank you!

File: plugins/wazo-snom/common/common.py (hunk 1 above)
Comment (BnLG) [reply plugins/wazo-snom/common/common.py]: done !

File: plugins/wazo-snom/common/common.py (hunk 2 above)
Comment (BnLG) [reply plugins/wazo-snom/common/common.py]: done !

File: plugins/wazo-snom/common/common.py (hunk 3 above)
Comment (BnLG) [reply plugins/wazo-snom/common/common.py:165]: Unfortunately no, as sorting by modification date (to get the most recent downloaded firmware) would then not be possible.

**Comment by BnLG** [comment]
Many thanks for your review @bloom1, I've answered your various points 👍

**Comment by bloom1** [comment]
Thank you for the reminder :+1:

**Comment by BnLG** [comment]
Thank you for your approval 👍

**Comment by BnLG** [comment]
@bloom1, as you approved this PR, let's then add the `mergeit` label ? :blush:
Thank you again !

**Comment by bloom1** [comment]
regate

**Comment by BnLG** [comment]
🎉 Many thanks 👍

File: plugins/wazo-snom/10.1.101.11/pkgs/pkgs.db [hunk 4]
```diff
@@ -75,60 +75,73 @@ version: 2.1.1
...
+[pkg_uxmc-fw]
+description: Firmware for Snom Extension USB Module UXMC D7C
+description_fr: Micrologiciel pour module d'extension USB Snom UXMC D7C
+version: 1.1.1
+files: uxmc-fw
+install: snom-fw
+
 [install_snom-fw]
 a-b: cp $FILE1 firmware/
+b-c: touch -c firmware/$FILE1
```
Comment (BnLG) [inline plugins/wazo-snom/10.1.101.11/pkgs/pkgs.db:87]: Mmmh seems like this makes the plugin fails to install (see last line of the stacktrace below) :
```
2022-07-04 16:54:29,721 [28336] (ERROR) (provd.app): Error while loading plugin wazo-snom-10.1.101.11
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/provd/app.py", line 824, in _pg_load
    self.pg_mgr.load(id, gen_cfg, spec_cfg)
  File "/usr/lib/python2.7/dist-packages/provd/plugins.py", line 1134, in load
    plugin = obj(self._app, plugin_dir, gen_cfg, spec_cfg)
  File "/var/lib/wazo-provd/plugins/wazo-snom-10.1.101.11/common.py", line 153, in __init__
    fetchfw_helper = FetchfwPluginHelper(plugin_dir, downloaders)
  File "/usr/lib/python2.7/dist-packages/provd/plugins.py", line 554, in __init__
    self._pkg_mgr = self._new_pkg_mgr(downloaders, filter_builder)
  File "/usr/lib/python2.7/dist-packages/provd/plugins.py", line 572, in _new_pkg_mgr
    pkg_builder)
  File "/usr/lib/python2.7/dist-packages/xivo_fetchfw/storage.py", line 386, in __init__
    self._load_pkgs()
  File "/usr/lib/python2.7/dist-packages/xivo_fetchfw/storage.py", line 393, in _load_pkgs
    self._pkgs = self._create_pkg(config, pkg_sections, remote_files, install_mgr_factories)
  File "/usr/lib/python2.7/dist-packages/xivo_fetchfw/storage.py", line 455, in _create_pkg
    pkg = self._pkg_builder.build_installable_pkg(config, section, pkg_id, remote_files, install_mgr_factories)
  File "/usr/lib/python2.7/dist-packages/xivo_fetchfw/storage.py", line 280, in build_installable_pkg
    pkg_install_mgr = install_mgr_factory.new_install_mgr(src_node, local_vars)
  File "/usr/lib/python2.7/dist-packages/xivo_fetchfw/storage.py", line 188, in new_install_mgr
    filter_obj = self._filter_builder.build_node(tokens)
  File "/usr/lib/python2.7/dist-packages/xivo_fetchfw/storage.py", line 144, in build_node
    raise ValueError("unknown node type %s" % type)
ValueError: unknown node type touch
```

Did I miss something to execute the `touch` system command ?
Thx 👍

**Pull request merged**